firmware/defmt-semihosting/src/lib.rs

```diff
@@ -1,56 +1,142 @@
-//! is writing to the buffer at a time. You must import a crate that provides a `critical-section` implementation
-//! suitable for the current target. See the `critical-section` README for details.
+//! This crate uses
+//! [`critical-section`](https://github.com/rust-embedded/critical-section) to
+//! ensure only one thread is writing to the buffer at a time. You must import a
+//! crate that provides a `critical-section` implementation suitable for the
+//! current target. See the `critical-section` README for details.
 //!
-//! For example, for single-core privileged-mode Cortex-M targets, you can add the following to your Cargo.toml.
+//! For example, for single-core privileged-mode Cortex-M targets, you can add
+//! the following to your Cargo.toml.
 //!
 //! ```toml
 //! [dependencies]
 //! cortex-m = { version = "0.7.6", features = ["critical-section-single-core"]}
 //! ```
 
 #![no_std]
-// nightly warns about static_mut_refs, but 1.76 (our MSRV) does not know about
-// that lint yet, so we ignore it it but also ignore if it is unknown
-#![allow(unknown_lints, static_mut_refs)]
 
-use core::sync::atomic::{AtomicBool, Ordering};
+use core::{
+    cell::UnsafeCell,
+    sync::atomic::{AtomicBool, Ordering},
+};
 
-use cortex_m_semihosting::hio;
+use semihosting::io::{Stdout, Write as _};
 
 #[defmt::global_logger]
 struct Logger;
 
-static TAKEN: AtomicBool = AtomicBool::new(false);
-static mut CS_RESTORE: critical_section::RestoreState = critical_section::RestoreState::invalid();
-static mut ENCODER: defmt::Encoder = defmt::Encoder::new();
+static SEMIHOSTING_ENCODER: SemihostingEncoder = SemihostingEncoder::new();
+
+struct SemihostingEncoder {
+    /// A boolean lock
+    ///
+    /// Is `true` when `acquire` has been called and we have exclusive access to
+    /// the rest of this structure.
+    taken: AtomicBool,
+    /// We need to remember this to exit a critical section
+    cs_restore: UnsafeCell<critical_section::RestoreState>,
+    /// A defmt::Encoder for encoding frames
+    encoder: UnsafeCell<defmt::Encoder>,
+    /// A semihosting handle for outputting encoded data
+    handle: UnsafeCell<Option<Stdout>>,
+}
 
-unsafe impl defmt::Logger for Logger {
-    fn acquire() {
+impl SemihostingEncoder {
+    /// Create a new semihosting-based defmt-encoder
+    const fn new() -> SemihostingEncoder {
+        SemihostingEncoder {
+            taken: AtomicBool::new(false),
+            cs_restore: UnsafeCell::new(critical_section::RestoreState::invalid()),
+            encoder: UnsafeCell::new(defmt::Encoder::new()),
+            handle: UnsafeCell::new(None),
+        }
+    }
+
+    /// Acquire the defmt encoder.
+    fn acquire(&self) {
         // safety: Must be paired with corresponding call to release(), see below
         let restore = unsafe { critical_section::acquire() };
 
-        if TAKEN.load(Ordering::Relaxed) {
+        // NB: You can re-enter critical sections but we need to make sure
```

Comment:
What does `NB` stand for? non-blocking?

Reply:
[*Nota Bene*](https://en.wiktionary.org/wiki/nota_bene)